src/deserialize.rs

```diff
@@ -0,0 +1,405 @@
+//! ```
+//!
+//! Alternatively, consider using the [Backup API](./backup/).
+
+use std::marker::PhantomData;
+use std::ptr::NonNull;
+use std::{alloc, fmt, iter, mem, ops, slice};
+
+use super::ffi;
+use crate::{Connection, DatabaseName, Result};
+
+impl Connection {
+    /// Disconnect from database and reopen as an in-memory database based on [`SerializedDb`].
+    pub fn deserialize(&mut self, db: DatabaseName<'_>, data: SerializedDb) -> Result<()> {
+        let result = unsafe {
+            self.deserialize_with_flags(
+                db,
+                &data,
+                data.cap,
+                DeserializeFlags::FREE_ON_CLOSE | DeserializeFlags::RESIZEABLE,
+            )
+        };
+        mem::forget(data);
+        result
+    }
+
+    /// Disconnect from database and reopen as an in-memory database based on the serialization data.
+    ///
+    /// # Safety
+    ///
+    /// The reference `data` must last for the lifetime of this connection.
+    /// `cap` must be the size of the allocation, and `cap >= data.len()`.
+    ///
+    /// If the data is not mutably borrowed, set [`DeserializeFlags::READ_ONLY`].
+    /// If SQLite allocated the memory, consider setting [`DeserializeFlags::FREE_ON_CLOSE`]
+    /// and/or [`DeserializeFlags::RESIZEABLE`].
+    ///
+    /// See the C Interface Specification [Deserialize a database](https://www.sqlite.org/c3ref/deserialize.html).
+    pub unsafe fn deserialize_with_flags(
+        &mut self,
+        db: DatabaseName<'_>,
+        data: &[u8],
+        cap: usize,
+        flags: DeserializeFlags,
+    ) -> Result<()> {
+        let path = &mut self.path;
+        let mut c = self.db.borrow_mut();
+        let db = db.to_cstring()?;
+        let rc = ffi::sqlite3_deserialize(
+            c.db(),
+            db.as_ptr(),
+            data.as_ptr() as *mut _,
+            data.len() as _,
+            cap as _,
+            flags.bits() as _,
+        );
+        c.decode_result(rc).map(|_| {
+            *path = Some(":memory:".into());
+        })
+    }
+
+    /// Returns the serialization of a database, or `None` when [`DatabaseName`] does not exist.
+    pub fn serialize(&self, db: DatabaseName<'_>) -> Result<Option<SerializedDb>> {
+        unsafe { self.serialize_with_flags(db, SerializeFlags::empty()) }
+    }
+
+    /// The serialization is the same sequence of bytes which would be written to disk
+    /// if the database where backed up to disk.
+    ///
+    /// Returns `Ok(None)` when [`DatabaseName`] does not exist or malloc/prepare fails.
+    ///
+    /// # Safety
+    ///
+    /// If [`SerializeFlags::NO_COPY`] is set, this returns a pointer to memory that SQLite is currently using
+    /// (or `Ok(None)` if this is not available).
+    /// In that case, the returned `SerializedDb` mutably borrows `Connection`,
+    /// [`std::mem::forget()`] one of them to prevent double free.
```

Comment:
I don't think this is a good thing for us to have in our API. There has to be a better way...

Reply:
OK, I made this function private and added `pub fn serialize_no_copy` as a safe alternative.